## Ejercicio 1:
1. Descarga de datos
•	Descarga los datasets para vinos tintos y blancos desde la siguiente URL: Wine Quality Dataset.
•	Asegúrate de que ambos datasets se encuentren disponibles en tu entorno de trabajo como archivos CSV.
•	Idealmente, intenta llevarlo a cabo de manera programática para no tener que hacer operaciones manualmente.




In [ ]:
import pandas as pd


In [51]:
df_redwine=pd.read_csv("C:/Users/CHOPPE/Desktop/master AI/Primer Trabajo de consolidacion MASTER/winequality-red.csv", sep=";")

In [52]:
df_redwine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [ ]:
#Comprobando que la creacion del dataframe haya sido exitosa
df_redwine.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   
5            7.4              0.66         0.00             1.8      0.075   
6            7.9              0.60         0.06             1.6      0.069   
7            7.3              0.65         0.00             1.2      0.065   
8            7.8              0.58         0.02             2.0      0.073   
9            7.5              0.50         0.36             6.1      0.071   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   
5                 13.0                  40.0   0.9978  3.51       0.56   
6                 15.0                  59.0   0.9964  3.30       0.46   
7                 15.0                  21.0   0.9946  3.39       0.47   
8                  9.0                  18.0   0.9968  3.36       0.57   
9                 17.0                 102.0   0.9978  3.35       0.80   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5  
5      9.4        5  
6      9.4        5  
7     10.0        7  
8      9.5        7  
9     10.5        5

In [ ]:
#Agregando columna de tipo de vino al data frame
df_redwine["tipo de vino"]="Red Wine"
df_redwine.head(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality tipo de vino  
0      9.4        5     Red Wine  
1      9.8        5     Red Wine  
2      9.8        5     Red Wine  
3      9.8        6     Red Wine  
4      9.4        5     Red Wine

In [ ]:
#Cargando el otro archivo csv y creando otro dataframe solo que esta vez con el fichero de vinos blancos
df_whitewine=pd.read_csv("C:/Users/CHOPPE/Desktop/master AI/Primer Trabajo de consolidacion MASTER/winequality-white.csv",sep=";")
df_whitewine.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   
5            8.1              0.28         0.40             6.9      0.050   
6            6.2              0.32         0.16             7.0      0.045   
7            7.0              0.27         0.36            20.7      0.045   
8            6.3              0.30         0.34             1.6      0.049   
9            8.1              0.22         0.43             1.5      0.044   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   
5                 30.0                  97.0   0.9951  3.26       0.44   
6                 30.0                 136.0   0.9949  3.18       0.47   
7                 45.0                 170.0   1.0010  3.00       0.45   
8                 14.0                 132.0   0.9940  3.30       0.49   
9                 28.0                 129.0   0.9938  3.22       0.45   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6  
5     10.1        6  
6      9.6        6  
7      8.8        6  
8      9.5        6  
9     11.0        6

In [ ]:
#Comprobando la creacion del dataframe haya sido exitosa y añadiendo la columna tipo de vino con su valor correspondiente
df_whitewine["tipo de vino"]="White wine"
df_whitewine.head(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality tipo de vino  
0      8.8        6   White wine  
1      9.5        6   White wine  
2     10.1        6   White wine  
3      9.9        6   White wine  
4      9.9        6   White wine


## Combinar los datos Ejercicio 2:
•	Usa Pandas para cargar ambos datasets en dataframes en memoria.
•	Combina los datos en un único dataframe añadiendo una columna adicional que indique el tipo de vino (red o white).
•	¿Cuántos registros tenemos? ¿Cuántas variables y de qué tipo?





In [ ]:
#unir ambos dataframes con las mismas columnas uno debajo de otro 
df_total = pd.concat([df_redwine, df_whitewine], ignore_index=True)
df_total.head(-5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00            1.90      0.076   
1               7.8             0.880         0.00            2.60      0.098   
2               7.8             0.760         0.04            2.30      0.092   
3              11.2             0.280         0.56            1.90      0.075   
4               7.4             0.700         0.00            1.90      0.076   
...             ...               ...          ...             ...        ...   
6487            6.8             0.220         0.36            1.20      0.052   
6488            4.9             0.235         0.27           11.75      0.030   
6489            6.1             0.340         0.29            2.20      0.036   
6490            5.7             0.210         0.32            0.90      0.038   
6491            6.5             0.230         0.38            1.30      0.032   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
6487                 38.0                 127.0  0.99330  3.04       0.54   
6488                 34.0                 118.0  0.99540  3.07       0.50   
6489                 25.0                 100.0  0.98938  3.06       0.44   
6490                 38.0                 121.0  0.99074  3.24       0.46   
6491                 29.0                 112.0  0.99298  3.29       0.54   

      alcohol  quality tipo de vino  
0         9.4        5     Red Wine  
1         9.8        5     Red Wine  
2         9.8        5     Red Wine  
3         9.8        6     Red Wine  
4         9.4        5     Red Wine  
...       ...      ...          ...  
6487      9.2        5   White wine  
6488      9.4        6   White wine  
6489     11.8        6   White wine  
6490     10.6        6   White wine  
6491      9.7        5   White wine  

[6492 rows x 13 columns]

In [ ]:
# Por medio del siguiente comando podemos identificar que tenemos 6497 datos por cada columna, en total son 12 columnas sin contar la columna de index. Ademas 
# Podemos ver que  casi todas las columnas númericas son de tipo float64, solo la columna 11 ("quality") es de tipo int64 y la columna 12 es de tipo object por ser un texto
df_total.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         6497 non-null   float64
 1   volatile acidity      6497 non-null   float64
 2   citric acid           6497 non-null   float64
 3   residual sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free sulfur dioxide   6497 non-null   float64
 6   total sulfur dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  tipo de vino          6497 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


In [139]:
numero_filas=df_total.shape[0]
numero_columnas=df_total.shape[1]
print( "el dataframe tiene ", numero_filas, "filas", "& tiene", numero_columnas, "columnas.", "Ademas casi todas las columnas son numericas tipo float64, solo la columna quality es de tipo int64")

el dataframe tiene  6497 filas & tiene 13 columnas. Ademas casi todas las columnas son numericas tipo float64, solo la columna quality es de tipo int64


## Ejercicio 3 Filtrar atípicos y manejar datos ausentes:
•	Realiza un análisis estadístico o inspección visual de cada columna numérica para identificar valores atípicos.
•	Usa Pandas para filtrar y eliminar los datos atípicos y los valores ausentes. Explica en tu entrega qué criterios utilizaste para identificar los atípicos.



In [63]:
# 1. Seleccionar solo columnas numéricas, generó una lista con el nombre de las columnas con valores numericos.
columnas_numericas = df_total.select_dtypes(include='number').columns

# 2. Creamos una máscara booleana que marca las filas que NO tienen valores atipicos
filtro_valores_no_atipicos = pd.Series([True] * len(df_total))

# 3. Recorremos cada columna numérica 
for i in columnas_numericas:
    Q1 = df_total[i].quantile(0.25)
    Q3 = df_total[i].quantile(0.75)
    IQR = Q3 - Q1
    limite_inferior = Q1 - 1.5 * IQR
    limite_superior = Q3 + 1.5 * IQR
    
    # 4. Creamos una máscara para esta columna (True si está dentro de los límites)
    mascara_columna = (df_total[i] >= limite_inferior) & (df_total[i] <= limite_superior)
    
    # 5. Actualizamos la máscara general (conserva solo las filas sin valores atipicos en ninguna columna)
    filtro_valores_no_atipicos &= mascara_columna

# 6. Filtramos el DataFrame lo hacemos con copy para no crear dependencias
df_total_limpio = df_total[filtro_valores_no_atipicos].copy()

# 7. Resultado
print(f"Filas originales: {len(df_total)}")
print(f"Filas después de eliminar valores atipicos: {len(df_total_limpio)}")

Filas originales: 6497
Filas después de eliminar valores atipicos: 4840


In [ ]:
#compruebo el resultado de mi dataframe limpio. Como muestra el comando mi dataframe limpio cuenta con 4840 filas. 
df_total_limpio.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4840 entries, 6 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4840 non-null   float64
 1   volatile acidity      4840 non-null   float64
 2   citric acid           4840 non-null   float64
 3   residual sugar        4840 non-null   float64
 4   chlorides             4840 non-null   float64
 5   free sulfur dioxide   4840 non-null   float64
 6   total sulfur dioxide  4840 non-null   float64
 7   density               4840 non-null   float64
 8   pH                    4840 non-null   float64
 9   sulphates             4840 non-null   float64
 10  alcohol               4840 non-null   float64
 11  quality               4840 non-null   int64  
 12  tipo de vino          4840 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 529.4+ KB


## Ejercicio 04 Almacenar los datos limpios en SQLite:
•	Usa SQLite para almacenar el dataframe limpio en una base de datos persistente.
•	Sigue la documentación oficial de SQLite: SQLite Python Documentation.




In [ ]:
import sqlite3
#Con el siguiente comando creao una base de datos en sqlLite 
conexion = sqlite3.connect("mi_base_de_datos.db")

In [75]:
#Con el siguiente comando creo una tabla dentro del mi base de datos y esa tabla queda guardada de forma persistente 
df_total_limpio.to_sql(name="tabla_vinos_limpia", con=conexion, if_exists='replace', index=False)

4840

In [ ]:
#Con el siguiente comando puedo hacer una consulta con la misma estructura del lenguaje SQL SELECT * FROM y convertir esa consulta en un dataframe de pandas.
prueba_consulta_sql=pd.read_sql("SELECT pH, alcohol, quality, density FROM tabla_vinos_limpia", conexion)
prueba_consulta_sql


pH  alcohol  quality  density
0     3.30      9.4        5  0.99640
1     3.35     10.5        5  0.99780
2     3.28      9.2        5  0.99590
3     3.35     10.5        5  0.99780
4     3.30     10.5        7  0.99690
...    ...      ...      ...      ...
4835  3.27     11.2        6  0.99114
4836  3.15      9.6        5  0.99490
4837  2.99      9.4        6  0.99254
4838  3.34     12.8        7  0.98869
4839  3.26     11.8        6  0.98941

[4840 rows x 4 columns]

## Ejercicio 05 Realizar 3 consultas en SQLite:
Realizar 3 consultas en SQLite Basándote en los datos y las columnas del dataset, realiza las siguientes consultas:
5.1	Consulta 1: ¿Cuál es el promedio de calidad (quality) por tipo de vino (type)?
5.2	Consulta 2: ¿Cuántos vinos tienen un nivel de alcohol superior a 10.5, agrupados por tipo?
5.3	Consulta 3: Obtén el conteo de vinos por nivel de acidez (fixed acidity) agrupados en rangos (por ejemplo, de 0-5, 5-10, 10-15).



In [105]:
promedio_calidad_tipo=pd.read_sql('SELECT quality, "tipo de vino" FROM tabla_vinos_limpia', conexion)
promedio_calidad_tipo.head(10)

quality tipo de vino
0        5     Red Wine
1        5     Red Wine
2        5     Red Wine
3        5     Red Wine
4        7     Red Wine
5        4     Red Wine
6        6     Red Wine
7        5     Red Wine
8        5     Red Wine
9        6     Red Wine

In [120]:
consulta='''
SELECT  
    "tipo de vino", 
    AVG(quality) AS promedio_calidad 
FROM 
    tabla_vinos_limpia 
GROUP BY 
    "tipo de vino"
'''

In [121]:
promedio_tipo_calidad=pd.read_sql(consulta, conexion)
promedio_tipo_calidad.head()

tipo de vino  promedio_calidad
0     Red Wine          5.656891
1   White wine          5.848244

In [136]:
promedio_tipo_calidad["promedio_calidad"].loc[1]

5.848244348244348